Fix #10 Implement read-rows-acceptance-test

ajaaym opened the pull request:
Fixes #10 Implement read-rows-acceptance-test 
- [x] Tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

ajaaym commented:
I signed it!

On Sat, Dec 16, 2017 at 9:51 PM, googlebot <notifications@github.com> wrote:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project. Before we can look at your
> pull request, you'll need to sign a Contributor License Agreement (CLA).
>
> 📝 *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address on your commit.
>    Check your existing CLA data <https://cla.developers.google.com/clas>
>    and verify that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If your company signed a CLA, they designated a Point of Contact who
>    decides which employees are authorized to participate. You may need to
>    contact the Point of Contact for your company and ask to be added to the
>    group of authorized contributors. If you don't know who your Point of
>    Contact is, direct the project maintainer to go/cla#troubleshoot. The email
>    used to register you as an authorized contributor must be the email used
>    for the Git commit.
>    - In order to pass this check, please resolve this problem and have
>    the pull request author add another comment and the bot will run again. If
>    the bot doesn't comment, it means it doesn't think anything has changed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/nodejs-bigtable/pull/22#issuecomment-352227793>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AglEnnEbHxF_6gCy3B15uhA4F1Uy5_MGks5tBIGdgaJpZM4REhwk>
> .
>

sduskis commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,217 @@
+ * @param {object} [options] Formatting options.
+ *
+ * @example
+ * Row.formatChunks_(chunks);
```
Please change this to `ChunkFormatter. formatChunks_`

ajaaym commented:
@kolodny Thanks for feedback. I updated PR as per pull request except about callback. Had discussion about callbacks vs returning array of rows with @sduskis  earlier and decided to use callback so to keep less number of live reference and also it helps to end stream earlier without parsing all chunks in a message.

I will create new PR with single commit log once it is approved to keep commit log clean.

sduskis commented:
Please don't create a new PR.  When the review is complete, we will "squash and merge" the PR, which will turn all commits into a single commit.

ajaaym commented:
Ok.

> On Dec 19, 2017, at 7:00 AM, Solomon Duskis <notifications@github.com> wrote:
> 
> Please don't create a new PR. When the review is complete, we will "squash and merge" the PR, which will turn all commits into a single commit.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sduskis commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+  this.family = row.data[chunk.familyName.value] = {};
+  const qualifierName = Mutation.convertFromBytes(chunk.qualifier.value);
+  this.qualifiers = this.family[qualifierName] = [];
+  this.qualifier = {
```
`this.qualifier` should probably be named `this.cellInProgress`.

ajaaym replied:
this.row, this.family and this.qualifier in the ChunkFormatter instance is in progress row, family & qualifier. Also there I named a method cellInProgress to process CELL_IN_PROGRESS state, It will be confusing with that name. If we want to make it explicit, I can rename state processing method to processNewRow, processRowInProgress, processCellInProgress & rename state holder to rowInProgress, familyInProgress & cellInProgress. Let me know.

sduskis commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+  }
+  this.qualifier.value =
+    this.qualifier.value + Mutation.convertFromBytes(chunk.value, options);
+  this.qualifier.size = 0;
```
Shouldn't `this.qualifier.size = chunk.size`?

ajaaym replied:
yes will update PR with this fix.

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+ */
+function ChunkFormatter() {
+  if (!(this instanceof ChunkFormatter)) {
+    return new ChunkFormatter();
```
Why is this block required?

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+ * @throws Error if there is any uncommitted row.
+ */
+ChunkFormatter.prototype.onStreamEnd = function() {
+  if (typeof this.row.key !== 'undefined') {
```
Use `isError`.

ajaaym replied:
updated

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+ * @param {string} text Error text
+ * @param {object} chunk chunk object to append to text
+ */
+ChunkFormatter.prototype.isError = function(condition, text, chunk) {
```
Can we reuse 'assert' which can be renamed to 'require' or use as is.

ajaaym replied:
I agree we need better name for this. I dont think we should use assert for this as assert is mainly for debugging/test case purpose.

chauhraj replied:
This is not necessarily true. Java/C++ both have the notion of asserts and they can exist very much in the code. Essentially, it is like preconditions. However, given that one can rename assert and call it either 'require', or 'precondition'.

kolodny replied:
I believe `invariant` would be the proper term https://github.com/facebook/react/blob/ea9714807b0d694e6f67d52ccf7aac93439b16f7/packages/react-native-renderer/src/ReactNativeComponentTree.js#L28

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+ * @private
+ */
+ChunkFormatter.prototype.reset = function() {
+  this.prevRowKey = '';
```
Shouldn't this be either undefined or null. It can be string or raw bytes depending on `decode` value.

ajaaym replied:
previous row key is assigned the row key of committed row which is always converted from bytes. Row key should never be blank even.

chauhraj replied:
Only reason, I was saying that if one look at [CellChunk::row-key] (https://github.com/googleapis/nodejs-bigtable/blob/master/protos/google/bigtable/v2/bigtable.proto#L105) datatype is bytes.

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+ */
+ChunkFormatter.prototype.isError = function(condition, text, chunk) {
+  if (condition) {
+    throw new Error(`${text}: ${chunk}`);
```
`${chunk}` will just convert to `[object Object]` To have a meaningful information about chunk, maybe use `${JSON.stringify(chunk)}` or something else.

ajaaym replied:
updated

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+    value: Mutation.convertFromBytes(chunk.value, options),
+    labels: chunk.labels,
+    timestamp: chunk.timestampMicros,
+    size: chunk.valueSize,
```
Is there any value in having size as part of the data? this.qualifier is the actual data in which user is interested in. chunk.valueSize is meant for the API to indicate when the Cell is complete. It helps API to pre-allocate a buffer to store the information and usually in bytes. Besides, everything is converted from bytes so it looses its relevance.

chauhraj commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,311 @@
+};
+
+module.exports = ChunkFormatter;
+module.exports.RowStateEnum = RowStateEnum;
```
Is there a reason to export RowStateEnum? Shouldn't this be private?

ajaaym replied:
I wanted to use it in test case, since its just based on number, I could have created new fake for enum in test case. I think its ok to export this, as this is immutable. But I will wait for more feedback if we shouldnt export this.

sduskis commented:
@ajaaym, what's the status on this?  Is this PR ready for review from @stephenplusplus yet?

ajaaym commented:
There are three open que where I need more feedback. 

1. Method name isError if anybody has suggestion for better name or anybody has preference to use assert instead?

2. Any issue exporting RowStateEnum. 

3. Size in output row. Previous code was sending out size in output. I agree with Rajiv that we dont need to output size. So should we keep it or remove it.

sduskis commented:
1. Maybe `checkCondition` or `validate` instead of `isError`?  There's a `[Preconditions](https://google.github.io/guava/releases/21.0/api/docs/com/google/common/base/Preconditions.html)` class in guava with this type of functionality

2. I don't see anything wrong.

3. The size of the output is necessary, and in the default cause (0) it's downright confusing.  I think it should be removed.

kolodny commented on src/table.js:
```diff
@@ -488,30 +489,45 @@ Table.prototype.createReadStream = function(options) {
+            },
+            (err, rowData) => {
+              if (stream._ended) {
+                throw new Error('stream ended');
```
It appears that you throw this error to "break" out of the loop in formatChunks, and catch the error on line 515. This doesn't feel like idiomatic javascript, can you instead return false in this loop and that will signal to `formatChunks` that you want to terminate early? Lodash follows a similar pattern https://lodash.com/docs/4.17.4#forEach

ajaaym commented:
@sduskis @kolodny @chauhraj Updated with following.
1. removed Size from output row.
2. Changed to stop processing of loop based on callback returns value.
3. Changed name isError to invariant

sduskis commented:
@stephenplusplus, we're ready for your review.

stephenplusplus commented on src/table.js:
```diff
@@ -488,30 +489,44 @@ Table.prototype.createReadStream = function(options) {
+                var row = self.row(rowData.key);
+                row.data = rowData.data;
+                throughStream.push(row);
+                return true;
```
No return necessary. This whole block could look like:

```js
if (stream._ended) {
  return;
}

var row = self.row(rowData.key);
row.data = rowData.data;
throughStream.push(row);
```

stephenplusplus replied:
After further digging, I see how the `return` value is used as the meaning of `shouldContinue` in ChunkFormatter. Can we be more direct about that, and just have a method like `chunkFormatter.exit()` (or `stopProcessing()`, etc). It's not clear that this function has any duties that work backwards into ChunkFormatter, and it would be easy to imagine a contributor or maintainer in the future getting rid of the `true`/`false` aspect in a refactor.

stephenplusplus commented on src/table.js:
```diff
@@ -488,30 +489,44 @@ Table.prototype.createReadStream = function(options) {
+        next();
+      },
+      function(callback) {
+        try {
```
It would be nice to remove the try/catch here as well, something more like:

```js
through.obj(
  function(data, enc, next) {
    // ...
    next();
  },
  chunkFormatter.onStreamEnd
);
```

`onStreamEnd` should accept a callback, and return an error if it ended prematurely.

stephenplusplus commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,319 @@
+  for (const chunk of chunks) {
+    switch (this.state) {
+      case RowStateEnum.NEW_ROW:
+        shouldContinue = this.newRow(chunk, options, callback);
```
Is it possible to use an instance variable to track if it should continue? This is not an obvious pattern-- passing a callback and receiving a value that doesn't match the purpose of the method.

stephenplusplus replied:
Pushing the `try/catch` here seems fine, like:

```js
for (const chunk of chunks) {
  switch (this.state) {
    case RowStateEnum.NEW_ROW:
      try {
        this.processNewRow(chunk, options);
      } catch (e) {
        callback(e);
      }
  }
}
```

stephenplusplus commented on src/chunkformatter.js:
```diff
@@ -0,0 +1,319 @@
+        shouldContinue = this.cellInProgress(chunk, options, callback);
+        break;
+    }
+    if (shouldContinue === false) {
```
Following my previous comment, this would be something like:

```js
if (this.wasProcessingExited) {
  break;
}
```

ajaaym commented:
After giving more thoughts to @stephenplusplus feedback, I think best way to implement this is stateful transform. So I renamed ChunkFormatter to ChunkTransformer and extended it with Streams Transform interface. Also added destroy method to make it compatible with pump & pumpify.

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,337 @@
+ * @param {callback} next callback will be called once data is processed, with error if any error in processing
+ */
+ChunkTransformer.prototype._transform = function(data, enc, next) {
+  try {
```
Doing the `try/catch` inside each data event seems heavy, and like perf may be negatively affected. Is it possible that if the data is invalid, the called method, i.e. `processNewRow()`, could just destroy the stream right there with the error?

ajaaym replied:
@stephenplusplus even after calling destroy right there, current execution path from point of error to for loop needs to be short circuit. I see below option to achieve this. Let me know if there is any other way.


 Use _error instance object and it will be initialized in constructor as below:

```javascript
  this._error = {
    hasError: false,
    errorObj: null,
  };
```
Invariant will be 
```javascript
  if (condition) {
    this._error.hasError = true;
    this._error.errorObj = new TransformError({
      message: text,
      chunk: chunk,
    });
  }
```
processNewRow will be
```javascript
ChunkTransformer.prototype.processNewRow = function(chunk) {
  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
  this.validateNewRow(chunk);
  if (this._error.hasError) {
    return;
  }
```
And transform will come out of for loop by checking on hasError
```javascript
ChunkTransformer.prototype._transform = function(data, enc, next) {
  for (const chunk of data.chunks) {
    if (this._error.hasError) {
      return next(this._error.errorObj);
    }
    switch (this.state) {
      case RowStateEnum.NEW_ROW:
        this.processNewRow(chunk, this.options);
        break;
      case RowStateEnum.ROW_IN_PROGRESS:
        this.processRowInProgress(chunk, this.options);
        break;
      case RowStateEnum.CELL_IN_PROGRESS:
        this.processCellInProgress(chunk, this.options);
        break;
    }
  }
  next();
};
```

In this case all conditions in validation method would still be checked only process* and transform/flush will be short circuited. If we make invariant return the condition we can use that to short circuit validation method as well like below.
```javascript
ChunkTransformer.prototype.invariant = function(condition, text, chunk) {
  if (condition) {
    this._error.hasError = true;
    this._error.errorObj = new TransformError({
      message: text,
      chunk: chunk,
    });
  }
  return condition;
};
```
validations method would be like this:
```javascript
ChunkTransformer.prototype.validateNewRow = function(chunk) {
  const row = this.row;
  const prevRowKey = this.prevRowKey;
  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
  return (
    this.invariant(
      typeof row.key !== 'undefined',
      'A new row cannot have existing state',
      chunk
    ) ||
    this.invariant(
      typeof chunk.rowKey === 'undefined' || chunk.rowKey === '',
      'A row key must be set',
      chunk
    ) ||
    this.invariant(chunk.resetRow, 'A new row cannot be reset', chunk) ||
    this.invariant(
      prevRowKey === newRowKey,
      'A commit happened but the same key followed',
      chunk
    ) ||
    this.invariant(!chunk.familyName, 'A family must be set', chunk) ||
    this.invariant(!chunk.qualifier, 'A column qualifier must be set', chunk) ||
    this.validateValueSizeAndCommitRow(chunk)
  );
};
```

if we make use of return value then we can use this return value all the way up to transform function to break for loop. 

My preference would be to use return value in case of error all the way up to transform and instance variable for error object. Let me know if you have any suggestion or better way of handling this.

stephenplusplus replied:
You should be able to re-use `this._destroyed` (set by `.destroy()`) to break out of the loop:

```js
ChunkTransformer.prototype._transform = function(data, enc, next) {
  for (const chunk of data.chunks) {
    if (this._destroyed) {
      return;
    }

    switch (this.state) {
      case RowStateEnum.NEW_ROW:
        this.processNewRow(chunk, this.options);
        break;
      case RowStateEnum.ROW_IN_PROGRESS:
        this.processRowInProgress(chunk, this.options);
        break;
      case RowStateEnum.CELL_IN_PROGRESS:
        this.processCellInProgress(chunk, this.options);
        break;
    }
  }

  next();
}
```

ajaaym replied:
Totally missed that. Thanks alot for pointing that out. I will update PR.

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,337 @@
+ * @param {string} text Error text
+ * @param {object} chunk chunk object to append to text
+ */
+ChunkTransformer.prototype.invariant = function(condition, text, chunk) {
```
When scanning the code, it's not clear that any call to `this.invariant()` could result in a thrown error. I would rather see something like this, so we aren't try/catching anywhere:

```diff
- this.invariant(condition, 'Error message.', value);
+ if (condition) {
+   this.destroy(this.generateError('Error message.', value));
+ }
```

You could use [create-error-class](https://github.com/floatdrop/create-error-class):

```js
const createErrorClass = require('create-error-class');

const TransformError = createErrorClass('TransformError', function (props) {
  this.message = `${props.message}: ${JSON.stringify(props.chunk)}`;
});

// ...

ChunkTransformer.prototype.validationMethod = function(chunk) {
  if (condition) {
    this.destroy(new TransformError({
      message: 'Error message.',
      chunk: chunk
    });
  }
};
```

ajaaym commented:
@stephenplusplus updated as per your feedback. Please review.

stephenplusplus commented on src/table.js:
```diff
@@ -488,30 +489,21 @@ Table.prototype.createReadStream = function(options) {
+      const throughStream = this;
+      const row = self.row(rowData.key);
+      row.data = rowData.data;
+      throughStream.push(row);
```
You can simplify this into:

```js
const row = self.row(rowData.key);
row.data = rowData.data;
next(null, row);
```

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+ChunkTransformer.prototype.processNewRow = function(chunk) {
+  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
+  this.validateNewRow(chunk);
+  if (this._destroyed) {
```
I think we can get rid of this. We do the same check in the transform function.

ajaaym replied:
validateNewRow has a check for existence of familyName and chunk.qualifier in chunk. which I later use it without checking.  

without check
```javascript
this.family = row.data[chunk.familyName.value] = {};
  const qualifierName = Mutation.convertFromBytes(chunk.qualifier.value);
```

Do you suggest to check for existence of that instead?
```javascript
  if (chunk.familyName && chunk.qualifier) {
    this.family = row.data[chunk.familyName.value] = {};
    const qualifierName = Mutation.convertFromBytes(chunk.qualifier.value);
    this.qualifiers = this.family[qualifierName] = [];
    this.qualifier = {
      value: Mutation.convertFromBytes(chunk.value, this.options),
      labels: chunk.labels,
      timestamp: chunk.timestampMicros,
    };
    this.qualifiers.push(this.qualifier);
    this.moveToNextState(chunk);
  }
```

stephenplusplus replied:
I just meant we should be able to get rid of the check if the stream was destroyed, as we do that same check from the transform function.

ajaaym replied:
I updated PR as per your latest feedback except this one to get rid of checks completely. I replaced the check this._destroyed with chunk.familyName && chunk.qualifier otherwise its throwing error bcz of this statement
```javascript
this.family = row.data[chunk.familyName.value] = {};
    const qualifierName = Mutation.convertFromBytes(chunk.qualifier.value);
```

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+ */
+ChunkTransformer.prototype.processRowInProgress = function(chunk) {
+  this.validateRowInProgress(chunk);
+  if (this._destroyed) {
```
Same as last comment; I believe we can get rid of this.

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+ */
+ChunkTransformer.prototype.processCellInProgress = function(chunk) {
+  this.validateCellInProgress(chunk);
+  if (this._destroyed) {
```
Same as last comment; I believe we can get rid of this.

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+  const row = this.row;
+  const prevRowKey = this.prevRowKey;
+  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
+  if (typeof row.key !== 'undefined') {
```
We could condense these a little bit:

```js
ChunkTransformer.prototype.validateNewRow = function(chunk) {
  const row = this.row;
  const prevRowKey = this.prevRowKey;
  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);

  var errorMessage;

  if (typeof row.key !== 'undefined') {
    errorMessage = 'A new row cannot have existing state';
  } else if (typeof chunk.rowKey === 'undefined' || chunk.rowKey === '') {
    errorMessage = 'A row key must be set';
  } else if (chunk.resetRow) {
    errorMessage = 'A new row cannot be reset';
  } else if (prevRowKey === newRowKey) {
    errorMessage = 'A commit happened but the same key followed';
  } else if (!chunk.familyName) {
    errorMessage = 'A family must be set';
  } else if (!chunk.qualifier) {
    errorMessage = 'A column qualifier must be set';
  }

  if (errorMessage) {
    this.destroy(new TransformError({message: errorMessage, chunk}));
    return;
  }

  this.validateValueSizeAndCommitRow(chunk);
};
```

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+ * @param {chunk} chunk chunk to validate for resetrow
+ */
+ChunkTransformer.prototype.validateResetRow = function(chunk) {
+  if (
```
Could we refactor this to something more like:

```js
ChunkTransformer.prototype.validateResetRow = function(chunk) {
  const containsData =
    chunk.rowKey ||
    chunk.familyName ||
    chunk.qualifier ||
    chunk.value ||
    chunk.timestampMicros > 0;

  if (chunk.resetRow && containsData) {
    this.destroy(
      new TransformError({
        message: 'A reset should have no data',
        chunk: chunk,
      })
    );
  }
};
```

stephenplusplus commented on src/chunktransformer.js:
```diff
@@ -0,0 +1,390 @@
+  if (chunk.resetRow) {
+    return this.reset();
+  }
+  this.qualifier.value =
```
We can save some space:

```js
this.qualifier.value += Mutation.convertFromBytes(chunk.value, this.options);
```

stephenplusplus commented:
@ajaaym @sduskis This LGTM. (Thanks for putting up with my comments). Is this ready to go from your end?

sduskis commented:
+1 from me.

sduskis commented:
@stephenplusplus, what does @ajaaym need to do to fix the CLA warning?

ajaaym commented:
Looks good to me

stephenplusplus commented:
No work needed for the CLA bot. It's a known limitation when two members collaborate, it can't validate both authors. But since we know we're both signed, we can ignore the warning.

Ping me whenever you'd like a new release 👍
